File: sunflower-core/src/main/java/org/tdf/sunflower/net/ChannelListener.java
package org.tdf.sunflower.net;

import org.tdf.sunflower.proto.Message;

public interface ChannelListener {
    ChannelListener NONE = new ChannelListener() {
        @Override
        public void onConnect(PeerImpl remote, Channel channel) {

        }

        @Override
        public void onMessage(Message message, Channel channel) {

        }

        @Override
        public void onError(Throwable throwable, Channel channel) {

        }

        @Override
        public void onClose(Channel channel) {

        }
    };

    // triggered when channel is open, only once in the life cycle of the channel
    void onConnect(PeerImpl remote, Channel channel);

    // when new message received
    void onMessage(Message message, Channel channel);

    // when error occurred
    void onError(Throwable throwable, Channel channel);

    // when the channel been closed
    void onClose(Channel channel);
}

File: sunflower-core/src/main/java/org/tdf/sunflower/net/ChannelListener.kt
package org.tdf.sunflower.net

import org.tdf.sunflower.proto.Message

interface ChannelListener {
    // triggered when channel is open, only once in the life cycle of the channel
    fun onConnect(remote: PeerImpl, channel: Channel)

    // when new message received
    fun onMessage(message: Message, channel: Channel, udp: Boolean = false)

    // when error occurred
    fun onError(throwable: Throwable, channel: Channel)

    // when the channel been closed
    fun onClose(channel: Channel)

    companion object {
        val NONE: ChannelListener = object : ChannelListener {
            override fun onConnect(remote: PeerImpl, channel: Channel) {}
            override fun onMessage(message: Message, channel: Channel, udp: Boolean) {}
            override fun onError(throwable: Throwable, channel: Channel) {}
            override fun onClose(channel: Channel) {}
        }
    }
}